lib/widgets/action_sheet.dart

```diff
@@ -402,3 +405,70 @@ class ShareButton extends MessageActionSheetMenuItemButton {
     }
   }
 }
+
+class MarkAsUnreadButton extends MessageActionSheetMenuItemButton {
+  MarkAsUnreadButton({
+    super.key,
+    required super.message,
+    required super.messageListContext,
+  }) : narrow = MessageListPage.ancestorOf(messageListContext).narrow;
+
+  final Narrow narrow;
+
+  @override IconData get icon => Icons.mark_chat_unread_outlined;
+
+  @override
+  String label(ZulipLocalizations zulipLocalizations) {
+    return zulipLocalizations.actionSheetMarkAsUnread;
+  }
+
+  @override void onPressed(BuildContext context) async {
+    Navigator.of(context).pop();
+    final zulipLocalizations = ZulipLocalizations.of(messageListContext);
+
+    try {
+      final store = PerAccountStoreWidget.of(messageListContext);
+      final connection = store.connection;
+      Anchor anchor = NumericAnchor(message.id);
+      final apiNarrow = narrow.apiEncode();
+
+      while (true) {
+        final result = await updateMessageFlagsForNarrow(connection,
+          anchor: anchor,
+          includeAnchor: true,
+          numBefore: 0,
+          numAfter: 1000,
+          narrow: apiNarrow,
+          op: UpdateMessageFlagsOp.remove,
+          flag: MessageFlag.read);
+
+        if (!context.mounted) return;
```

Comment:
Looks like this is misused here. A similar issue from before: https://github.com/zulip/zulip-flutter/pull/779#discussion_r1676347611.